interpreter/src/lib.rs

```diff
@@ -37,44 +39,117 @@ impl State {
-        self.memory.insert(format!("{thread}.{x}"), value);
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.memory.insert(format!("{thread}.{x}"), value);
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    buffer.insert(format!("{thread}.{x}"), value);
+                } else {
+                    //
+                }
+            }
+        }
+    }
+
+    pub fn read(&mut self, x: &str, thread: &str) -> u32 {
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.memory.get(x).copied().unwrap_or(0)
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    if buffer.contains_key(x) {
+                        buffer.get(x).copied().unwrap_or(0)
+                    } else {
+                        self.memory.get(x).copied().unwrap_or(0)
+                    }
+                    
+                } else {
+                    // this happens when main thread executes
+                    self.memory.get(x).copied().unwrap_or(0)
+                }
+            }
+        }
+       
     }
 
-    pub fn read(&self, x: &str) -> u32 {
-        self.memory.get(x).copied().unwrap_or(0)
+    pub fn read_local(&mut self, thread: &str, x: &str) -> u32 {
+        match self.memory_model {
+            MemoryModel::Sc => {
+                self.read(format!("{thread}.{x}").as_str(), thread)
+            }
+            MemoryModel::Tso => {
+                if let Some(buffer) = self.write_buffers.get_mut(thread) {
+                    if buffer.contains_key(format!("{thread}.{x}").as_str()) {
+                        buffer.get(format!("{thread}.{x}").as_str()).copied().unwrap_or(0)
+                    } else {
+                        self.read(format!("{thread}.{x}").as_str(), thread)
+                    }
+                } else {
+                    self.read(format!("{thread}.{x}").as_str(), thread)
+                }
+            }
+        }
     }
 
-    pub fn read_local(&self, thread: &str, x: &str) -> u32 {
-        self.read(format!("{thread}.{x}").as_str())
+    /// Flushes a random thread-local variable to global variables for a specified thread.
+    pub fn flush_random_write_buffer(&mut self, thread_name: &str) -> bool {
+        if let Some(buffer) = self.write_buffers.get_mut(thread_name) {
+            let keys: Vec<String> = buffer.keys().cloned().collect();
+            if keys.is_empty() {
+                return false; 
+            }
+
+            let mut rng = thread_rng();
+            let random_index = rng.gen_range(0..keys.len());
+            let random_key = &keys[random_index];
+
+            if let Some(value) = buffer.remove(random_key) {
+                self.memory.insert(random_key.clone(), value);
+                return true; 
+            }
```

Comment:
This is incorrect. The write buffer should follow FIFO ordering to preserve WW ordering in TSO.